src/compiler/program.ts

```diff
@@ -619,20 +619,38 @@ namespace ts {
      * in cases when we know upfront that all load attempts will fail (because containing folder does not exists) however we still need to record all failed lookup locations.
      */
     function loadModuleFromFile(candidate: string, extensions: string[], failedLookupLocation: string[], onlyRecordFailures: boolean, state: ModuleResolutionState): string {
+        // First try to keep/add an extension: importing "./foo.ts" can be matched by a file "./foo.ts", and "./foo" by "./foo.d.ts"
+        const keepOrAddExtension = loadModuleFromFileWorker(candidate, extensions, failedLookupLocation, onlyRecordFailures, state);
+        if (keepOrAddExtension) {

```

Comment:
Nit: It says `keepOrAddExtension`, but that name implies that this is a boolean used for making a decision rather than the module itself. Consider `unstrippedResolvedName` or something like that.